caliban_toolbox/aws_functions.py

```diff
@@ -0,0 +1,166 @@
+        self._seen_so_far = 0
+        self._lock = threading.Lock()
+
+    def __call__(self, bytes_amount):
+        with self._lock:
+            self._seen_so_far += bytes_amount
+            percentage = (self._seen_so_far / self._size) * 100
+            sys.stdout.write(
+                "\r%s  %s / %s  (%.2f%%)" % (
+                    self._filename, self._seen_so_far, self._size,
+                    percentage))
+            sys.stdout.flush()
+
+
+def connect_aws():
+    AWS_ACCESS_KEY_ID = getpass('What is your AWS access key id? ')
+    AWS_SECRET_ACCESS_KEY = getpass('What is your AWS secret access key id? ')
+
+    session = boto3.Session(aws_access_key_id=AWS_ACCESS_KEY_ID,
+                            aws_secret_access_key=AWS_SECRET_ACCESS_KEY)
+    print('Connected to AWS')
+    s3 = session.client('s3')
+
+    return s3
+
+
+def aws_upload_files(aws_folder, stage, upload_folder, pixel_only, label_only, rgb_mode):
+    """Uploads files to AWS bucket for use in Figure 8
+
+    Args:
+        aws_folder: folder where uploaded files will be stored
+        stage: specifies stage in pipeline for jobs requiring multiple rounds of annotation
+        upload_folder: path to folder containing files that will be uploaded
+        pixel_only: boolean flag to set pixel_only mode
+        label_only: boolean flag to set label_only mode
+        rgb_mode: boolean flag to set rgb_mode
+    """
+
+    s3 = connect_aws()
+
+    # load the images from specified folder but not the json log file
+    files_to_upload = list_npzs_folder(upload_folder)
+
+    filename_list = []
+
+    # change slashes separating nested folders to underscores for URL generation
+    subfolders = re.split('/', aws_folder)
+    subfolders = '__'.join(subfolders)
+
+    # optional flags to add to URL
+    optional_flags = np.any([pixel_only, label_only, rgb_mode])
+    if optional_flags:
+        optional_url = "?"
+        if pixel_only:
+            optional_url += "&pixel_only=true"
+        if label_only:
+            optional_url += "&label_only=true"
+        if rgb_mode:
+            optional_url += "&rgb=true"
+
+    # upload images
+    for img in files_to_upload:
+
+        # full path to image
+        img_path = os.path.join(upload_folder, img)
+
+        # destination path
+        img_key = os.path.join(aws_folder, stage, img)
+
+        # upload
+        s3.upload_file(img_path, 'caliban-input', img_key, Callback=ProgressPercentage(img_path),
+                       ExtraArgs={'ACL': 'public-read', 'Metadata': {'source_path': img_path}})
+        print('\n')
+
+        url = "https://caliban.deepcell.org/{0}__{1}__{2}__{3}__{4}".format('caliban-input',
+                                                                            'caliban-output',
+                                                                            subfolders, stage, img)
+
+        if optional_flags:
+            url += optional_url
```

Comment:
I would prefer to always use the same quote char (so single quotes for the URL).

If you opt to try the urlencoding, you can just update `url` to be:

```python
url = 'https://caliban.deepcell.org/{0}__{1}__{2}__{3}__{4}?{5}'.format(
    'caliban-input',
    'caliban-output',
    subfolders, stage, img,
    url_encoded_dict
)
```